### PR TITLE
Build new bad-ass emock framework.

### PR DESCRIPTION
**Pull request by marshall-mcmullen**
emock is a mocking framework for bash. It integrates well into the rest of ebash and etest in particular to make it very easy to mock out real system binaries and replace them with mock instances which do something else. It is very flexible in the behavior of the mocked function utilizing the many provided option flags.                                             

The simplest invocation of emock is to use the default of function mocking. In this mode, you simply supply the name of the binary you wish to mock, such as:

`emock "dmidecode"`
                                                                                                                              
 This will create and export a new function called 'dmidecode' that can be invoked instead of the real 'dmidecode'  binary at '/usr/bin/dmidecode'. This also creates a function named 'dmidecode_real' which can be invoked to get access to the real underlying dmidecod binary at "/usr/sbin/dmidecode".                                                             
                                                                                                                               
By default, this mock function will simply return '0' and produce no stdout or stderr. This behavior can be customized using the options --return-code, --stdout, and --stderr.                                                                     
                                                                                                                               
Mocking a real binary with a simplex name like this is the simplest, but doesn't always work. In particular, if at the call site you call it with the fully-qualified path to the binary, as in 'usr/sbin/dmidecode', then our mocked function won't be called. In this scenario, you need to  mock it with the fully qualified path just as you would invoke it at the call site. For example:                                                                                                      
                                                                                                                               
 `emock "/usr/sbin/dmidecode"`
                                                                                                                              
 Just as before, this will create and export a new function named "/usr/sbin/dmidecode" which will be called in place of the real dmidecode binary. It will also create a "/usr/sbin/dmidecode_real" function which will point to the real binary in case you need to call it instead.                                                                                         
                                                                                                                               
emock tracks various metadata about mocked binaries for easier testability. This includes the number of times a mock is called, as well as the arguments (newline delimieted arg array), exit code, stdout, and stderr for each invocation. By default this is created in a local hidden directory named '.emock' and there will be a directory beneath that for each       
mock:                                                                                                                        
   
```
 .emock/dmidecode/called                                                                                                
 .emock/dmidecode/0/{args,exit,stdout,stderr}                                                                           
 .emock/dmidecode/1/{args,exit,stdout,stderr}                                                                           
 ```

There are lots of super awesome utility functions, such as:

* emock
* eunmock
* emock_called
* emock_stdout
* emock_stderr
* emock_args
* assert_emock_callecd
* assert_emock_stdout
* assert_emock_stderr
* assert_emock_called_with

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -745,6 +745,138 @@ or file contents that might later make it a lot easier to figure out what went w
...
+else. It is very flexible in the behavior of the mocked function utilizing the many provided option flags.
+
+The simplest invocation of emock is to use the default of function mocking. In this mode, you simply supply the name
+of the binary you wish to mock, such as:
+
+    # emock "dmidecode"
+
+This will create and export a new function called 'dmidecode' that can be invoked instead of the real 'dmidecode'
+binary at '/usr/bin/dmidecode'. This also creates a function named 'dmidecode_real' which can be invoked to get access
+to the real underlying dmidecod binary at "/usr/sbin/dmidecode".
```
Comment (thecodenomad) [inline README.md:761]: Typo around dmidecode, but also, holy crap that's awesome to mock out the binary+path.

File: README.md [hunk 2]
```diff
@@ -745,6 +745,138 @@ or file contents that might later make it a lot easier to figure out what went w
...
+
+Inside the statedir, the argument array is stored with each argument fully quoted so that whitespace encapsulated
+arguments preserve whitespace. To convert this back into an array, the best thing to do is to use array_init:
+
+`array_init args "$(emock_args func)`
+
+Alternatively, the helper idiom assert_emock_called_with is an extremely useful way to validate the arguments passed
+into a particular invocation.
+
+#### assert_emock_called
```
Comment (thecodenomad) [inline README.md:839]: :heart_eyes:

File: README.md [hunk 3]
```diff
@@ -745,6 +745,138 @@ or file contents that might later make it a lot easier to figure out what went w
...
+assert_emock_stderr is used to assert that a particular invocation of a mock produced the expected standard error.
+
+For example:
+
+```
+assert_emock_stderr "func" 0 "This is the expected standard error for call #0"
+assert_emock_stderr "func" 1 "This is the expected standard error for call #1"
+```
+
+#### assert_emock_called_with
```
Comment (thecodenomad) [inline README.md:867]: :heart_eyes:

File: share/emock.sh [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        ":stdout      o                     | What standard output should be returned by the mock."                    \
+        ":stderr      e                     | What standard error should be returned by the mock."                     \
+        ":statedir=.emock                   | This directory is used to track state about mocked binaries. This will
+                                              hold metadata information such as the number of times the mock was called
+                                              as well as the exit code, stdout, and stderr for each invocation."       \
+        "+delete      d                     | Delete existing mock state inside statedir from prior mock invocations." \
+        "name                               | Name of the binary to mock (e.g. dmidecode or /usr/sbin/dmidecode). This
+                                              must match the calling convention at the call site."                     \
+        "?body                              | This allows fine-grained control over the body of the mocked function that
```
Comment (thecodenomad) [inline share/emock.sh:66]: AWESOME!

File: share/emock.sh [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    )
+
+    assert_eq "${times}" "$(opt_forward emock_called statedir -- ${name})"
+}
+
+opt_usage assert_emock_stdout <<'END'
+assert_emock_stdout is used to assert that a particular invocation of a mock produced the expected standard output.
+For example:
+
+    # assert_emock_stdout "func" 0 "This is the expected standard output for call #0"
```
Comment (thecodenomad) [inline share/emock.sh:317]: Nice!

**Review by thecodenomad — APPROVED** [review]
:smiling_imp: This is awesome stuff, I can't see, or think, of anything else I'd add at the moment. I can't wait to start using it :grinning:

**Comment by ericpe** [comment]
WOW!  this is nice.  This concept used to be difficult for people to wrap their heads around

File: share/emock.sh [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+zero-based (again, just like an array).
+
+So if this has never been called, then emock_called will return 0, and there will be no on-disk state directory. The
+first time you call it, emock_called will return 1, and there will be a ${statedir}/0 directory storing the state files
+for that invocation.
+END
+emock_called()
+{
+    $(opt_parse \
+        ":statedir=.emock                   | This directory is used to track state about mocked binaries. This will
```
Comment (ericpe) [inline share/emock.sh:191]: Something to consider: have the statedir be something like ".emock-${pid}" so you can parallelize stuff using emock

Comment (thecodenomad) [reply share/emock.sh:191]: Ooo, I like this idea!! great suggestion @ericpe!

Comment (marshall-mcmullen) [reply share/emock.sh]: OH man, that is a great idea. I've implemented this!! Thanks for the suggestion @ericpe

**Pull request merged**